Add new feature for supporting CMIS as an external file store

ianwallen opened the pull request:
Add new file store for supporting CMIS supported application.

https://en.wikipedia.org/wiki/Content_Management_Interoperability_Services

By adding support external content management systems it allows for the following benefits.
- Support rich content and functionality of a content management services.
- Central storage allowing better support for clustering (future version of GN)
- Similar to cloud storage such as S3 however it does not require cloud services to run.  

Setup was tested using Open Text and ,  Alfresco.

## Changes in this PR
- Add support for CMIS to store resources on the CMIS supported application
- Update resource dialog box to support versioning and management link (if enabled in configuration) 
![image](https://user-images.githubusercontent.com/1868233/97224502-f7db6b80-17af-11eb-81e1-fb5d5cc56037.png)
- Update the config-spring-geonetwork.xml configuration so that file store selection so that it can be done at runtime without requiring custom build.  This allow for the selection of the default file store (FileSystem), CMIS or S3.

## Example Alfresco

Alfresco is the easiest to setup

- Quickly setup and Alfresco installation using Virtual Box and the following image.
https://bitnami.com/stack/alfresco/virtual-machine
- Create a user in alfresco that will be the owner of the content.
- Create the CMIS_BASE_REPOSITORY_PATH (suggest: geonetwork) in alfresco and grant the user created above write access.

Next you can set the configuration files or use env configuration.

### Example (configuration files)
- Set CMIS as the file store
https://github.com/geonetwork/core-geonetwork/compare/master...ianwallen:cmis_support?expand=1#diff-89b263990ea6093c0c5e5d1269eb327e65959228d9da5d8676f4556ca55d1d5fR193-R204
- Set CMIS configuration (See config-cmis.xml)

### Example (using env variables)
Enable CMIS store
```
       GEONETWORK_STORE_TYPE=cmis
```
CMIS configuration
```
       CMIS_REPOSITORY_ID=-default-
       CMIS_USERNAME=username
       CMIS_PASSWORD=password
       CMIS_SERVICES_BASE_URL=http://localhost:8080/alfresco
       CMIS_BASE_REPOSITORY_PATH=geonetwork
       CMIS_BINDING_TYPE=browser
       CMIS_BROWSER_URL=/api/-default-/public/cmis/versions/1.1/browser
```
To see version number in File store
```
       CMIS_VERSIONING_ENABLED=true
```
To see external resource management link in File store
[Configuration option](https://github.com/ianwallen/core-geonetwork/blob/cmis_support/core/src/main/java/org/fao/geonet/api/records/attachments/CMISStore.java#L467-L484)
```
       CMIS_EXTERNAL_RESOURCE_MANAGEMENT_URL=http://localhost:8080/share/page/document-details?nodeRef=workspace://SpacesStore/{cmisobjectid}
```
Optional external resource management configuration
```
       CMIS_EXTERNAL_RESOURCE_MANAGEMENT_WINDOW_PARAMETERS=toolbar=0,width=600,height=600
       CMIS_EXTERNAL_RESOURCE_MANAGEMENT_MODAL_ENABLED=true
```

# Known issue
- When working with the workflow the working copy will only copy the last version of the document to the working copy so working copies will also start with version 1 even though the approved copy is at a higher version.  This may cause confusion.
Versions could be copied as well however this could cause a large performance hit. (see the following https://stackoverflow.com/questions/48669716/cmis-alfresco-copy-all-versions-of-a-document)
A new issue will be created to look into this issue after this PR is approved.

- All logos are not loaded on initial load. Only the default gn logo is loaded.

jodygarnett commented:
This should be good, as little interaction with other systems.

Surfacing version form content management system is an interesting capability.

jodygarnett commented on core/pom.xml:
```diff
@@ -355,6 +355,24 @@
+    <dependency>
+      <groupId>org.apache.chemistry.opencmis</groupId>
+      <artifactId>chemistry-opencmis-client-api</artifactId>
+      <version>1.1.0</version>
```
manage id root pom dependency management?

ianwallen replied:
@jodygarnett 
I'm assuming that you mean copy the pom dependency to the root pom.xml and then remove the version information from the core/pom.xml
As this implementation is similar to the S3 implementation I copied the same method that was used by the following.

```
 <dependency>
      <groupId>com.amazonaws</groupId>
      <artifactId>aws-java-sdk-s3</artifactId>
      <version>1.11.618</version>
    </dependency>
```

I also noticed that there are lots of other dependencies with versions in the core/pom.xml

If this is a new requirement for the GN project then I suggest that we do this cleanup as a new PR.
Or maybe I'm misunderstanding what you are suggesting?

jodygarnett replied:
I understand best practice, is not always current practice.

josegar74 commented:
@ianwallen, doing tests with Alfresco I noticed some issues:

1) Deleting a metadata with resources I get an error 500, the request returns:

```
{"message":"CmisRuntimeException","code":"runtime_exception","description":"500"}
```

No error in the log.

2) Logo management only shows the default logo, seem the other logos in the filesystem are not loaded by default in Alfresco. Not sure if related to this PR or it's a issue of using other storage options like S3 also.

3) Uploading a new logo, it appears in Alfresco `images/harvesting` folder, but it's not displayed in the Logo Management page

josegar74 commented on core/src/main/java/org/fao/geonet/api/records/attachments/CMISStore.java:
```diff
@@ -0,0 +1,618 @@
+            versionValue = version;
+        }
+
+        MetadataResource.ExternalResourceManagementProperties externalResourceManagementProperties = getExternalResourceManagementProperties(context, metadataId, metadataUuid, visibility, resourceId, filename, version);
```
Review long lines like this one, to add breaklines.

ianwallen replied:
I'm not sure what the max length should be?  I changed my settings to wrap at 200 and it wrapped that line.

If 200 is the suggested length then I suggest adding `max_line_length = 200` to the `.editorconfig` file.

josegar74 commented on core/src/main/java/org/fao/geonet/resources/CMISConfiguration.java:
```diff
@@ -0,0 +1,443 @@
+            }
+        }
+
+        parameters.put(SessionParameter.REPOSITORY_ID, repositoryId);
```
Should be relevant for `null` and throw an Exception?

ianwallen replied:
I added a null check

josegar74 commented on core/src/main/java/org/fao/geonet/resources/CMISConfiguration.java:
```diff
@@ -0,0 +1,443 @@
+     */
+    private String getServiceUrl(String baseUrl, String serviceUrl, String defaultServicePathInfo) {
+        // If no service url was not supplied then lets default to the base url plus the default path info
+        if (serviceUrl == null || serviceUrl.isEmpty()) {
```
Maybe use `StringUtils.isEmpty`

ianwallen replied:
Changed

josegar74 commented on core/src/main/java/org/fao/geonet/resources/CMISResources.java:
```diff
@@ -0,0 +1,491 @@
+package org.fao.geonet.resources;
```
Add file header

ianwallen replied:
Done

josegar74 commented on core/src/main/java/org/fao/geonet/resources/CMISConfiguration.java:
```diff
@@ -0,0 +1,443 @@
+package org.fao.geonet.resources;
```
Add file header

ianwallen replied:
Done

josegar74 commented on core/src/main/java/org/fao/geonet/resources/CMISResources.java:
```diff
@@ -0,0 +1,491 @@
+            }
+
+            if (webappCopy == null) {
+                webappCopy = appPath.resolve(filename);  // TODO: this won't work...
```
Is this something to solve?

ianwallen replied:
Copied from S3.  Code was updated in latest changes.

josegar74 commented on core/src/main/java/org/fao/geonet/resources/CMISResources.java:
```diff
@@ -0,0 +1,491 @@
+                        }
+                    }
+                } catch (CmisObjectNotFoundException e) {
+                    if (writeOnClose) {
```
Should not be moved to a `finally` block?

ianwallen replied:
I don't think a try finally block will work here as it is only removing the path if the object does not exists otherwise it is loading it from the cmis repository.

josegar74 commented on core/src/main/java/org/fao/geonet/resources/CMISResources.java:
```diff
@@ -0,0 +1,491 @@
+                    }
+                }
+            }
+            java.nio.file.Files.delete(path);
```
Should be added in `try / finally` block?

ianwallen replied:
It was taken from S3 implementation which did not hava a try catch finally block to handle the errors.
I added it

ianwallen commented on core/src/main/resources/config-store/config-cmis.xml:
```diff
@@ -0,0 +1,84 @@
+
+<!-- Open Text sample -->
+<!--<bean id="cmisconfiguration" class="org.fao.geonet.resources.CMISConfiguration">-->
+<!--    <property name="repositoryName" value="Enterprise"/>-->
```
Open text has several repositories so we had to specify a repository name.  The default main one seems to be Enterprise.
Alfresco only has one repository.  So if it was not supplied, it was picking the first one.  I checked again and I noticed that the default repository has repository id = `-default-`
I also realized that it may be a bad practice to pick the first repository.  If a second is ever supported then it could end up picking the wrong repository on startup.  So I removed the logic to pick the first repo and changed the example so that it uses  `CMIS_REPOSITORY_ID=-default-`

ianwallen commented on core/src/main/java/org/fao/geonet/api/records/attachments/CMISStore.java:
```diff
@@ -0,0 +1,618 @@
+        PathMatcher matcher =
+                FileSystems.getDefault().getPathMatcher("glob:" + filter);
+
+        OperationContext operationContext = CMISConfiguration.getClient().createOperationContext();
```
removed

ianwallen commented:
@josegar74 

> About versioning:
> 
> * `Update resource dialog box to support versioning and management link (if enabled in configuration)`, I don't see the icon, just the version, I setup this property: `export CMIS_VERSIONING_ENABLED=true`. Is it required any additional configuration?
> * I updated the document version in Alfresco, but not sure if possible to do it in Geonetwork uploading the file again?

To see the management link you need the CMIS_EXTERNAL_RESOURCE_MANAGEMENT variables set.  I have updated the documentation to include a sample for Alfresco.
Note: The sample provided is a link to the alfresco editor. In our case we created a small custom application for the popup which allows for a few custom properties to be set.  

Regarding the versioning - in my testing I was simply uploading the file on top of itself several time and the version would increase each time.  So yes, it should be possible to upload the file on top of the existing one.

ianwallen commented:
Note: if testing with workflow then ensure that the following changes are applied. https://github.com/geonetwork/core-geonetwork/pull/5219

ianwallen commented:
@josegar74 
> 1. Deleting a metadata with resources I get an error 500, the request returns:
> 
> ```
> {"message":"CmisRuntimeException","code":"runtime_exception","description":"500"}
> ```
> 

I applied a bunch of fixes - tested deleting record and I did not get a CmisRuntimeException error.

> No error in the log.
> 
> 1. Logo management only shows the default logo, seem the other logos in the filesystem are not loaded by default in Alfresco. Not sure if related to this PR or it's a issue of using other storage options like S3 also.
> 2. Uploading a new logo, it appears in Alfresco `images/harvesting` folder, but it's not displayed in the Logo Management page

There were a couple bugs with the resources - I had not tested it much. I had copied most of the code from the s3 implementation.  There are a couple bugs that I have fixed which should be applied to the S3 code as well.  But I did not want to change this as I cannot test S3.

It should work now. Please retest.

As for the icons loaded at startup - I only get the default icon.  Looks like there others are not loaded if not existing.  This is a separate issue and will look into fixing this in another PR.

ianwallen commented:
@josegar74 
I believe I fixed all the issues that you raised so could you please review again.

Note that there are 2 other PR that may affect the testing. It would be appreciated if you could review those 2 issues as well.

#5219 
#5223

Thank you

josegar74 commented:
Tested the changes and it's fixed the delete of metadata and now I get the version icon, that works fine.

I notice another issue. Clicking the `lock` icon in the resources panel, that should move the resource to `public` / `private` folder throws a `Bad request`:

![resources-type](https://user-images.githubusercontent.com/1695003/100599367-3957bd00-3300-11eb-963c-3b4493372ae1.png)

josegar74 commented:
The previous issue is fixed, but I noticed this parameter in the resource:

<pre>
<gmd:linkage>
<gmd:URL>http://localhost:8080/geonetwork/srv/api/records/6f061a57-5222-4ccb-a644-bab93951f1e1/attachments/gn-user-meeting-202006-GeoNetwork4-All-about-search.pdf?approved=false</gmd:URL>
</gmd:linkage>
</pre>

I approved / published the metadata and the published version has the parameter `approved=false` If the workflow is disabled the resources also get `approved=false` parameter.

Tested with CMIS and also with local filesystem as data directory.

ianwallen commented:
@josegar74
I have reverted some changes and applied the fix in another way and it seems to work better.

But then I noticed that the working copy resource was not linked correctly.  As you point out in  https://github.com/geonetwork/core-geonetwork/pull/5223#pullrequestreview-540737703

So I added to new issue #5226 for the issue that you reported.

ianwallen commented:
@josegar74  Could you update the milestone for this PR. I'm assuming that it will be targeting the 3.12 release?
Thank you
